<html>
  <head>
    <link rel = "stylesheet" href = "default.css">
  </head>
  <body>
<h1>Transopson Insertion Finder (TIF) <font size = "4"> - An efficient detector for insertions of transposable element from NGS data</font></h1>
TIF detectes transposition of transpososable elements from sequence data of NGS. This page explains how to use TIF, using the transposition detection of the rice endogenous retrotransposon <i>Tos17</i> as an example.
<h2><i>Tos17</i> in the last century</h2>
<p><img src = "tos17.gif" align = "left">
<i>Tos17</i> is an endogenous retrotransposon of rice. <i>Tos17</i> is active only in cultured cells of <i>japonica</i> rice cultivar Nipponbare (<a href = "https://doi.org/10.1073/pnas.93.15.7783" target = "_blank">Hirochika <i>et al.</i> 1996</a>). Taking advantage of this property, we created 50,000 lines of mutant panels from rice individuals regenerated from cultured cells (<a href = "https://doi.org/10.1105/tpc.012559" target = "_blank">Miyao <i>et al.</i> 2003</a>). Insertion sites of <i>Tos17</i> is determined by <a href = "https://doi.org/10.5511/plantbiotechnology.15.253" target = "_blank">the Suppression PCR method and TAIL-PCR method</a>, and is published in the <a href = "https://tos.nias.affrc.go.jp/" target = "_blank">mutant panel database</a>. <i>Tos17</i> insertion mutant lines have been distributed to researchers around the world and is <a href = "https://scholar.google.com/scholar?hl=ja&as_sdt=0%2C5&q=tos17&btnG=" target = "_blank">useful for functional analysis of genes</a>.
<p>Before the development of NGS technology, primer set of a primer having a <i>Tos17</i> sequence near the end of <i>Tos17</i> and  a primer with a special modification at the end, or a short primer with the elaborate reaction cycles was used to amplify the DNA fragment containing the flanking sequence of <i>Tos17</i> insertion point. Amplified fragments were separated by agarose gel electrophoresis, and the DNA obtained from the gel block containing the fragment was used as a template to determine the flanking sequence of <i>Tos17</i> insertion point by the Sanger method.
<p>We tried to determine insertion positions of <i>Tos17</i> for all 50,000 lines. It was costly and required several years. Furthermore, amplification of flanking sequences depends on recognition sites of restriction enzyme or by-chance primer annealing, making it difficult to cover all insertion sites.
<h2><i>Tos17</i> in this century with NGS</h2>
<p>NGS technology has made it possible to obtain the entire genome sequence of rice individuals. The junction created by <i>Tos17</i> insertion should also be included in the large number of short reads. Sequences having a junction with <i>Tos17</i> can be easily selected by searching the sequences at both ends of <i>Tos17</i>. It is known that a 5-base duplication occurs at the target site during the insertion of <i>Tos17</i>. <a href = "https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-71/figures/1" target = "_blank">The insertion site of <i>Tos17</i> can be known by selecting a set having the same 5 base target site duplication (TSD) in the 5'-end and the 3'-end-flankng sequence of <i>Tos17</i></a>. TIF selects short reads containing 5'-end and 3'-end sequence of transposable element, makes set of 5'-end and the 3'-end-flankng sequence having same TSD sequence, and then output inserted position on the reference genome by mapping pair of flanking seqeunces.
<h2>Detection of <i>Tos17</i> insertion from NGS</h2>
<p>Once the NGS sequence data of the <i>Tos17</i> insertion mutant individual is obtained, the insertion site can be easily obtained by the <a href = "https://github.com/akiomiyao/tif" target = "_blank">TIF</a> program. A computer running unix (<a href = "https://ubuntu.com" target = "_blank">ubuntu</a> is good), NGS data, and the rice reference genome sequence (<a href = "https://rapdb.dna.affrc.go.jp/download/irgsp1.html" target = "_blank">IRGSP1.0</a>) are required.
<p>Following is an example.
<p>To obtain short read sequence, fastq-dump is required. <a href = "https://akiomiyao.github.io/ped/sratoolkit/index.html" target = "_blank">Link to fastq-dump setup</a> 
<pre>% git clone https://github.com/akiomiyao/tif.git
$ cd git
$ mkdir ttm5
$ mkdir ttm5/read
$ cd ttm5/read
$ fastq-dump --split-files SRR556174   # Fastq-dump takes long time.
$ fastq-dump --split-files SRR556175

$ cd git
$ wget https://rapdb.dna.affrc.go.jp/download/archive/irgsp1/IRGSP-1.0_genome.fasta.gz
$ perl tif.pl IRGSP-1.0_genome.fasta.gz  ttm5 TGTTAAATATATATACA TTGCAAGTTAGTTAAGA
</pre>
<p>'<a href = "https://doi.org/10.1186/1471-2229-10-175" target = "_blank">ttm5</a>' is one of the <i>Tos17</i> insertion line (line name: ND6834). The NGS data of ttm5 was acquired for <a href = "https://doi.org/10.1093/pcp/pcr172" target = "_blank">the analysis of somaclonal variation</a>. ND6834 was harvested as a regenerated individual in October 1998. The individual was regenerated from liquid culture in N6 medium containing 2,4-dichlorophenoxyacetic acid at a concentration of 1 mg/L until March 1998 from the Nipponbare calus induced in October 1997. The individual analyzed by NGS is the M<sub>4</sub> generation, two generations after the harvested seeds. <a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR556174" target = "_blank">SRR556174</a> and <a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR556175" target = "_blank">SRR556175</a> are run data of ttm5 by Illumina Hiseq2000 sequencing system.
<p>$ perl tif.pl<br>
without argument returns how to use tif.pl.<br>
'$' at the head of line indicates prompt of terminal window.
<p>$ perl tif.pl reference target_directory 5'-end_of_TE_sequence 3'-end_of_TE_sequence<br>
For analysis of rice, IRGSP-1.0_genome.fasta.gz is the reference genome sequence.<br>
<i>Tos17</i> 5'-end (Head) sequence: TGTTAAATATATATACA<br>
<i>Tos17</i> 3'-end (Tail) sequence: TTGCAAGTTAGTTAAGA<br>
<p>Result of tif.pl is saved in target (ttm5) directory.

<h2>Contact</h2>
MIYAO Akio, Ph.D (<a href = "mailto:miyao@affrc.go.jp">miyao@affrc.go.jp</a>) (Family name is MIYAO)<br>
Institute of Crop Science / <a href = "http://www.naro.affrc.go.jp/english/index.html" target = "_blank">NARO</a><br>
2-1-2, Kannondai, Tsukuba, Ibaraki, 305-8518, Japan
  </body>
</html>
