<html>
  <head>
    <link rel = "stylesheet" href = "default.css">
  </head>
  <body>
<h1>Transopson Insertion Finder (TIF) <font size = "4"> - An efficient detector for insertions of transposable element from NGS data</font></h1>
TIF detectes transposition of transpososable elements from sequence data of NGS. This page explains how to use TIF, using the transposition detection of the rice endogenous retrotransposon <i>Tos17</i> as an example.<br>
Code: <a href = "https://github.com/akiomiyao/tif" target = "_blank">https://github.com/akiomiyao/tif</a><br>
Paper: <a href = "https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-71" target = "_blank">Nakagome M, Solovieva E, Takahashi A, Yasue H, Hirochika H, Miyao A (2014) Transposon Insertion Finder (TIF): a novel program for detection of de novo transpositions of transposable elements. BMC Bioinformatics 5:71.</a>
<h2><i>Tos17</i> across the century</h2>
<p><img src = "tos17.gif" align = "left">
<i>Tos17</i> is an endogenous retrotransposon of rice. <i>Tos17</i> is active only in cultured cells of <i>japonica</i> rice cultivar Nipponbare (<a href = "https://doi.org/10.1073/pnas.93.15.7783" target = "_blank">Hirochika <i>et al.</i> 1996</a>). Taking advantage of this property, we created about 50,000 lines of mutant panels from rice individuals regenerated from cultured cells (<a href = "https://doi.org/10.1105/tpc.012559" target = "_blank">Miyao <i>et al.</i> 2003</a>). Insertion sites of <i>Tos17</i> is determined by <a href = "https://doi.org/10.5511/plantbiotechnology.15.253" target = "_blank">the Suppression PCR method and TAIL-PCR method</a>, and is published in the <a href = "https://tos.nias.affrc.go.jp/" target = "_blank">mutant panel database</a>. <i>Tos17</i> insertion mutant lines have been distributed to researchers around the world and is <a href = "https://scholar.google.com/scholar?hl=ja&as_sdt=0%2C5&q=tos17&btnG=" target = "_blank">useful for functional analysis of genes</a>.
<p>Before the development of NGS technology, primer set of a primer having a <i>Tos17</i> sequence near the end of <i>Tos17</i> and  a primer with a special modification at the end, or a short primer with the elaborate reaction cycles was used to amplify the DNA fragment containing the flanking sequence of <i>Tos17</i> insertion point. Amplified fragments were separated by agarose gel electrophoresis, and the DNA obtained from the gel block containing the fragment was used as a template to determine the flanking sequence of <i>Tos17</i> insertion point by the Sanger method.
<p>We tried to determine insertion positions of <i>Tos17</i> for all 50,000 lines. It was costly and required several years. Furthermore, amplification of flanking sequences depends on recognition sites of restriction enzyme or by-chance primer annealing, making it difficult to cover all insertion sites.
<h2><i>Tos17</i> in this century with NGS</h2>
<p>NGS technology has made it possible to obtain the entire genome sequence of rice individuals. The junction created by <i>Tos17</i> insertion should also be included in the large number of short reads. Sequences having a junction with <i>Tos17</i> can be easily selected by searching the sequences at both ends of <i>Tos17</i>. It is known that a 5-base duplication occurs at the target site during the insertion of <i>Tos17</i>. <a href = "https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-71/figures/1" target = "_blank">The insertion site of <i>Tos17</i> can be known by selecting a set having the same 5 base target site duplication (TSD) in the 5'-end and the 3'-end-flankng sequence of <i>Tos17</i></a>. TIF selects short reads containing 5'-end and 3'-end sequence of transposable element, makes set of 5'-end and the 3'-end-flankng sequence having same TSD sequence, and then output inserted position on the reference genome by mapping pair of flanking seqeunces.
<h2>Detection of <i>Tos17</i> insertion from NGS data</h2>
<p>Once the NGS sequence data of the <i>Tos17</i> insertion mutant individual is obtained, the insertion site can be easily obtained by the <a href = "https://github.com/akiomiyao/tif" target = "_blank">TIF</a> program. A computer running unix (<a href = "https://ubuntu.com" target = "_blank">ubuntu</a> is good), NGS data, and the rice reference genome sequence (<a href = "https://rapdb.dna.affrc.go.jp/download/irgsp1.html" target = "_blank">IRGSP1.0</a>) are required.
<p>Following is an example.
<p>To obtain short read sequence, fastq-dump is required. <a href = "https://akiomiyao.github.io/ped/sratoolkit/index.html" target = "_blank">Link to fastq-dump setup</a><br>
git clone and wget reference is required at first analysis.
<pre>$ git clone https://github.com/akiomiyao/tif.git
$ cd git
$ mkdir ttm5
$ mkdir ttm5/read
$ cd ttm5/read
$ fastq-dump --split-files SRR556174   # Fastq-dump takes long time.
$ fastq-dump --split-files SRR556175

$ cd git
$ wget https://rapdb.dna.affrc.go.jp/download/archive/irgsp1/IRGSP-1.0_genome.fasta.gz
$ perl tif.pl IRGSP-1.0_genome.fasta.gz  ttm5 TGTTAAATATATATACA TTGCAAGTTAGTTAAGA
</pre>
<p>'<a href = "https://doi.org/10.1186/1471-2229-10-175" target = "_blank">ttm5</a>' is one of the <i>Tos17</i> insertion line (line name: ND6834). The NGS data of ttm5 was acquired for <a href = "https://doi.org/10.1093/pcp/pcr172" target = "_blank">the analysis of somaclonal variation</a>. ND6834 was harvested as a regenerated individual in October 1998. The individual was regenerated from liquid culture in N6 medium containing 2,4-dichlorophenoxyacetic acid at a concentration of 1 mg/L until March 1998 from the Nipponbare calus induced in October 1997. The individual analyzed by NGS is the M<sub>4</sub> generation, two generations after the harvested seeds. <a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR556174" target = "_blank">SRR556174</a> and <a href = "https://www.ncbi.nlm.nih.gov/sra/?term=SRR556175" target = "_blank">SRR556175</a> are run data of ttm5 by Illumina Hiseq2000 sequencing system.
<p>without argument returns how to use tif.pl.<br>
$ perl tif.pl<br>
'$' at the head of line indicates prompt of terminal window.<br>
<p>$ perl tif.pl reference target_directory 5'-end_of_TE_sequence 3'-end_of_TE_sequence<br>
For analysis of rice, IRGSP-1.0_genome.fasta.gz is the reference genome sequence.<br>
<i>Tos17</i> 5'-end (Head) sequence: TGTTAAATATATATACA<br>
<i>Tos17</i> 3'-end (Tail) sequence: TTGCAAGTTAGTTAAGA<br>
<p>Result of tif.pl is saved in target (ttm5) directory.
<pre>
miyao@s2:~/tif/ttm5$ ls -la
total  432
drwxr-xr-x 4 miyao miyao  45056 Jun 27 09:25 ./
drwxrwxr-x 6 miyao miyao   4096 Jun 27 09:22 ../
-rw-rw-r-- 1 miyao miyao   2096 Jun 27 09:25 TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA.vcf
drwxrwxr-x 2 miyao miyao   4096 Jun 27 09:25 child/
-rw-rw-r-- 1 miyao miyao 264216 Jun 25 11:56 grep.TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA
-rw-rw-r-- 1 miyao miyao  45127 Jun 27 09:25 log.TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA
drwxrwxr-x 2 miyao miyao   4096 Jun 15 15:50 read/
-rw-rw-r-- 1 miyao miyao   2368 Jun 27 09:25 result.TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA
-rw-rw-r-- 1 miyao miyao  55348 Jun 27 09:24 selected.TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA
</pre>
<p>head_seq.tail_seq.vcf is the result with vcf format.<br>
grep.head_seq.tail_seq is the list of short reads containing head_seq or tail_seq.<br>
selected.head_seq.tail_seq is the list of short reads which are not present in the reference genome sequence.<br>
log.head_seq.tail_seq is the log of run.<br>
result.head_seq.tail_seq is the result of tif analysis.<br><br>
<b>Result with vcf format</b>
<pre>miyao@s2:~/tif/ttm5$ cat TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA.vcf 
##fileformat=VCFv4.3
##fileDate=20210627
##source=&lt;PROGRAM=tif.pl,target=ttm5,reference=IRGSP-1.0_genome.fasta&gt;
##INFO=&lt;ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant"&gt;
##INFO=&lt;ID=MEINFO,Number=9,Type=String,Description="Movile element info of the form ME_HEAD_SEQ,ME_TAIL_SEQ,JUNCTION_POS_OF_HEAD,JUNCTION_POS_OF_TAIL,TSD_SIZE,TSD_SEQUENCE,DIRECTION,COUNT_OF_READS_WITH_JUNCTION_OF_HEAD,COUNT_OF_READS_WITH_JUNCTION_OF_TAIL"&gt;
##ALT=&lt;ID=INS:ME,Type=String,Description="Insertion of a mobile element"&gt;
##created=&lt;TIMESTAMP="2021-06-27 09:25:50 +0900"&gt;
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO
chr01   34453641        .       G       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,34453641,34453645,5,CTTTG,reverse,2,18;SVTYPE=INS
chr02   1004765 .       C       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,1004765,1004769,5,ATACC,reverse,1,9;SVTYPE=INS
chr02   31596632        .       T       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,31596632,31596628,5,CTAAT,forward,4,11;SVTYPE=INS
chr03   741222  .       C       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,741222,741226,5,GCAGC,reverse,1,12;SVTYPE=INS
chr03   8304674 .       A       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,8304674,8304678,5,GAATA,reverse,2,15;SVTYPE=INS
chr06   24967877        .       T       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,24967877,24967881,5,TGCAT,reverse,3,19;SVTYPE=INS
chr07   20064395        .       T       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,20064395,20064391,5,CTTAT,forward,4,24;SVTYPE=INS
chr07   20080556        .       T       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,20080556,20080552,5,CTTAT,forward,4,24;SVTYPE=INS
chr09   12970614        .       C       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,12970614,12970618,5,CATGC,reverse,1,19;SVTYPE=INS
chr10   19069889        .       G       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,19069889,19069885,5,ACTTG,forward,2,8;SVTYPE=INS
chr10   21583058        .       T       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,21583058,21583054,5,CTTAT,forward,2,7;SVTYPE=INS
chr12   2155895 .       C       &lt;INS:ME&gt;        .       .       MEINFO=TGTTAAATATATATACA,TTGCAAGTTAGTTAAGA,2155895,2155899,5,GGAAC,reverse,4,12;SVTYPE=INS
</pre>
<b>Table 1.</b> Table of <i>Tos17</i> transposed positions in ttm5. TIF output, 'result.TGTTAAATATATATACA.TTGCAAGTTAGTTAAGA' separated with TAB is converted to the html table. TSDs in flanking sequence are indicated with red.
<table border>
<tr><th>Chr</th><th>Junction of head</th><th>Junction of tail</th><th>TSD size</th><th>TSD</th><th>Direction</th><th>Head flanking sequence (HFS)</th><th>Tail flanking sequence (TFS)</th><th>Detected number of HFS</th><th>Detected number of TFS</th></tr>
<tr><td>chr01</td><td align = "right">34453641</td><td align = "right">34453645</td><td align = "center">5</td><td>CTTTG</td><td align = "center">reverse</td><td>AGCAAGAAGTATCTCATCAGAGCCACCTTCACTTATGGAAACTACGATGGGCTCAACTCGTCAGAGAAGGGTT<font color = "red">CTTTG</font></td><td><font color = "red">CTTTG</font>TTTATCTTTGGACTCCATATCGGTGTCAACTTCTGGACGACGGTAAACTTGACAAAGTGGGATCCATCGA</td><td align = "right">2</td><td align = "right">18</td>
<tr><td>chr02</td><td align = "right">1004765</td><td align = "right">1004769</td><td align = "center">5</td><td>ATACC</td><td align = "center">reverse</td><td>AAATTTGAGATATACCATCCTGACCCACATGTCATTGACTCATGTGGGTTCTACATGTCATTGAT<font color = "red">ATACC</font></td><td><font color = "red">ATACC</font>AATGGCATATCTCAAACTTTGCAAAATTATAATGGCATGGCTCCAATTTACCCTTTT</td><td align = "right">1</td><td align = "right">9</td>
<tr><td>chr02</td><td align = "right">31596632</td><td align = "right">31596628</td><td align = "center">5</td><td>CTAAT</td><td align = "center">forward</td><td>CAAACACCGTGTAATTTAGAGGGATGCTATATAACGGAATCTCGTTCAGACGGGATGATACCTACGAGGTATAAGTT<font color = "red">CTAAT</font></td><td><font color = "red">CTAAT</font>ACCTGTGAGAGTATCGACTTCAGATACCGGTGAGTATTGAATGATTCATGCGAGCTATCA</td><td align = "right">4</td><td align = "right">11</td>
<tr><td>chr03</td><td align = "right">741222</td><td align = "right">741226</td><td align = "center">5</td><td>GCAGC</td><td align = "center">reverse</td><td>AACCGGCTGTCGGATGTGTTTACTGTGTTTGCTGGACATTTGGCAATCAACGTTGGTGGTG<font color = "red">GCAGC</font></td><td><font color = "red">GCAGC</font>CAGGGTTCCATCCCTTCCAGCTCAGGTGGATCGAGCTGTAGAATCTCTGGTTCACCATAAGTTAG</td><td align = "right">1</td><td align = "right">12</td>
<tr><td>chr03</td><td align = "right">8304674</td><td align = "right">8304678</td><td align = "center">5</td><td>GAATA</td><td align = "center">reverse</td><td>GCATAAAGAGTGAATCTCATTGAAGGAACAAAATGAACATTTCAATTTTTTTCAGCAGTCTCGGTAATCAAAGTTAT<font color = "red">GAATA</font></td><td><font color = "red">GAATA</font>TACCATTGCTCTATAGCACCTCTCTCTGGCAAAATATTAGCATTTTGTCTCAAGACCTCCTTATGACACTATAT</td><td align = "right">2</td><td align = "right">15</td>
<tr><td>chr06</td><td align = "right">24967877</td><td align = "right">24967881</td><td align = "center">5</td><td>TGCAT</td><td align = "center">reverse</td><td>GGAAATGATAGAGTTGCCAGAATTGGCAGAATGGTCTTCTGTGGACTGTTTTTTTCCTGCTCTTCTTGAGGTG<font color = "red">TGCAT</font></td><td><font color = "red">TGCAT</font>CAGAGGATGCCCCAAGCTGAAACAATTGCCACCAGTTGTTCTCCCACCTGTAAGAATGAGTATATATGTATCTACC</td><td align = "right">3</td><td align = "right">19</td>
<tr><td>chr07</td><td align = "right">20064395</td><td align = "right">20064391</td><td align = "center">5</td><td>CTTAT</td><td align = "center">forward</td><td>TCGCTCACAGTGGAGGACATTGTGGACAATGCAATCGTACTCCTGACTGCCGGTTATGGGAATTCGGCTGTT<font color = "red">CTTAT</font></td><td><font color = "red">CTTAT</font>CACCTTCTTGCTCCGGTACCTAGCTAATGATCCAGACATCCTTGGCAAGATAACCGAGGGTGAGCACCTAGTACTCT</td><td align = "right">4</td><td align = "right">24</td>
<tr><td>chr07</td><td align = "right">20080556</td><td align = "right">20080552</td><td align = "center">5</td><td>CTTAT</td><td align = "center">forward</td><td>TCGCTCACAGTGGAGGACATTGTGGACAATGCAATCGTACTCCTGACTGCCGGTTATGGGAATTCGGCTGTT<font color = "red">CTTAT</font></td><td><font color = "red">CTTAT</font>CACCTTCTTGCTCCGGTACCTAGCTAATGATCCAGACATCCTTGGCAAGATAACCGAGGGTGAGCACCTAGTACTCT</td><td align = "right">4</td><td align = "right">24</td>
<tr><td>chr09</td><td align = "right">12970614</td><td align = "right">12970618</td><td align = "center">5</td><td>CATGC</td><td align = "center">reverse</td><td>ACATTTGTATTAATTTCTATGGCCTATGGGGTTAGTGTTTGGACTGTCAAGGTGTGAGGCATCTTCTCAGTGACTGTG<font color = "red">CATGC</font></td><td><font color = "red">CATGC</font>ATGGTGGGGTGTAGTAGTAGTTTATCAGGACCAAAGAAAAGAACAAACTGCAAGCTAGCTGTAAGCTGCAAAGTAT</td><td align = "right">1</td><td align = "right">19</td>
<tr><td>chr10</td><td align = "right">19069889</td><td align = "right">19069885</td><td align = "center">5</td><td>ACTTG</td><td align = "center">forward</td><td>CAAAGTTAAAAGATTTTCTCACCAATAACCTATGGTATCTACCATTATATATCATCAGCAC<font color = "red">ACTTG</font></td><td><font color = "red">ACTTG</font>TACTCTGTGAAAATAGATTTCAGTAGTATGGTTATTGACACTGACAGAGCTGCTTGCTCAATACAATGTGTTTC</td><td align = "right">2</td><td align = "right">8</td>
<tr><td>chr10</td><td align = "right">21583058</td><td align = "right">21583054</td><td align = "center">5</td><td>CTTAT</td><td align = "center">forward</td><td>ATGGAGGTCGAGGTGGAGTTGCCAACAGCTCATTCTGTAAATCTATCTTCAAGTCCGAGCTCGAGGTA<font color = "red">CTTAT</font></td><td><font color = "red">CTTAT</font>CATATCCTCAAATTAGCTAGCCCCCCTATCGAGAATAGCTTAGCATATTTGAGGACCTGTTGGG</td><td align = "right">2</td><td align = "right">7</td>
<tr><td>chr12</td><td align = "right">2155895</td><td align = "right">2155899</td><td align = "center">5</td><td>GGAAC</td><td align = "center">reverse</td><td>TTAAATTTAGCTGAGATAATCTTGTTTTGCGACAAAGTTGATGTATTATAGAATGATTCCTTGAATTGCAGCTT<font color = "red">GGAAC</font></td><td><font color = "red">GGAAC</font>AATTTGGCTTTGTGTAGCACATAAATAAAACATCTTACTATTTAATGTGAACCAGCCAAACTTTGACCCTTGATGT</td><td align = "right">4</td><td align = "right">12</td>
</table>
<h2>Book</h2>
<a href = "https://www.springer.com/gp/book/9781071611333" target = "_blank">Plant transposable Elements, Methods in Molecular Biology, Cho, Jungnam (Ed.), 2021, Springer.</a>
<h2>Contact</h2>
MIYAO Akio, Ph.D (<a href = "mailto:miyao@affrc.go.jp">miyao@affrc.go.jp</a>) (Family name is MIYAO)<br>
Institute of Crop Science / <a href = "http://www.naro.affrc.go.jp/english/index.html" target = "_blank">NARO</a><br>
2-1-2, Kannondai, Tsukuba, Ibaraki, 305-8518, Japan
  </body>
</html>
